runtime/Java/src/org/antlr/v4/runtime/IncrementalParser.java

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2019 The ANTLR Project. All rights reserved.
+ * Licensed under the BSD-3-Clause license. See LICENSE file in the project root for license information.
+ */
+package org.antlr.v4.runtime;
+
+import org.antlr.v4.runtime.misc.Interval;
+import org.antlr.v4.runtime.tree.ErrorNode;
+import org.antlr.v4.runtime.tree.ParseTreeListener;
+import org.antlr.v4.runtime.tree.TerminalNode;
+
+/**
+ * Incremental parser implementation
+ * <p>
+ * There are only two differences between this parser and the underlying regular
+ * Parser - guard rules and min/max tracking
+ * <p>
+ * The guard rule API is used in incremental mode to know when a rule context
+ * can be reused. It looks for token changes in the bounds of the rule.
+ * <p>
+ * The min/max tracking is used to track how far ahead/behind the parser looked
+ * to correctly detect whether a token change can affect a parser rule in the future (IE when
+ * handed to the guard rule of the next parse)
+ *
+ * @note See IncrementalParsing.md for more details on the theory behind this.
+ * In order to make this easier in code generation, we use the parse
+ * listener interface to do most of our work.
+ */
+public abstract class IncrementalParser extends Parser implements ParseTreeListener {
+	// Current parser epoch. Incremented every time a new incremental parser is
+	// created.
+	private static int _PARSER_EPOCH = 0;
```

Comment:
This should be an `AtomicInteger` to avoid race conditions when multiple threads instantiate this class.